### PR TITLE
new post: How to use agent-side information on the puppetserver

### PR DESCRIPTION
**Pull request by DavidS**

### CODE REVIEW

File: _posts/2020-03-18-how-to-use-agent-information-on-puppetserver.md [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+  - custom facts
+  - custom resources
+  - deferred
+---
+
+This question was posted on the [Puppet Community Slack](https://slack.puppet.com/) yesterday:
+
+> Is there any way like the below in puppet code we can use:
+> ```
+> $var=‘run comman locally on agent, if $?=0 echo “present”, else echo “absent”’
```
Comment (hpcprofessional) [inline _posts/2020-03-18-how-to-use-agent-information-on-puppetserver.md]: s/comman/command/ ?

File: _posts/2020-03-18-how-to-use-agent-information-on-puppetserver.md [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+> if $var == present {
+>   # my code
+> } else {
+>   # something else
+> }
+> ```
+(by user [IlovPuppet](https://app.slack.com/team/URGFPBS0N), edited for presentation)
+
+From my answer there, I wrote up a more detailed exposition here.
```
Comment (hpcprofessional) [inline _posts/2020-03-18-how-to-use-agent-information-on-puppetserver.md]: I suggest that at this point in the blog, the "Why" of this is not yet answered. I would suggest adding a sentence or two explaining the kinds of things it enables, or why someone would want to use agent-side information.

To see what I mean,  How is IlovPuppet's problem not solved quickly and directly by a custom/external fact? Why do we need something other than a fact?

File: _posts/2020-03-18-how-to-use-agent-information-on-puppetserver.md [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+  notify { example :
+    message => $d
+  }
+}
+```
+(from [voxpupuli/puppet-vault_lookup](https://github.com/voxpupuli/puppet-vault_lookup#usage))
+
+**Use it ...**
+
+* for values only needed locally on a few agents.
```
Comment (hpcprofessional) [inline _posts/2020-03-18-how-to-use-agent-information-on-puppetserver.md]: This is a little unclear. Might this wording be better?
* for values specific to one or a few agents, for which managing via Hiera would be cumbersome

File: _posts/2020-03-18-how-to-use-agent-information-on-puppetserver.md [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+
+**Use it for ...**
+
+* things that can be modeled as a puppet resource.
+* fine-grained control.
+* full integration into the puppet tooling.
+
+**Don't use it if ...**
+
+* you do not want to code ruby.
```
Comment (hpcprofessional) [inline _posts/2020-03-18-how-to-use-agent-information-on-puppetserver.md]: Although it's technically true that writing Types/Providers using the resource API still requires Ruby, it is much easier than it used to be. I realize that technically the doc link below mentions it, but I'd suggest calling out the PDK and Resource API specifically as making it easier than the olden days...

File: _posts/2020-03-18-how-to-use-agent-information-on-puppetserver.md [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+* a simpler solution (see above) works.
+
+**Further Reading**
+
+* [Custom Resources](https://puppet.com/docs/puppet/latest/custom_resources.html): main documentation on writing custom resources.
+* [HUE hands on lab](https://github.com/puppetlabs/puppet-resource_api/tree/master/docs): hands-on lab for a custom remote resource interacting with a API.
```
Comment (hpcprofessional) [inline _posts/2020-03-18-how-to-use-agent-information-on-puppetserver.md]: I suggest an "In Conclusion" paragraph or similar might be appropriate to tie everything together, if only to thank IluvPuppet for their blog-generating question.

File: _posts/2020-03-18-how-to-use-agent-information-on-puppetserver.md (hunk 1 above)
Comment (DavidS) [reply _posts/2020-03-18-how-to-use-agent-information-on-puppetserver.md]: yes

File: _posts/2020-03-18-how-to-use-agent-information-on-puppetserver.md (hunk 2 above)
Comment (DavidS) [reply _posts/2020-03-18-how-to-use-agent-information-on-puppetserver.md]: I'll add a few words here. Part of the problem is that I don't know much about the motivations of people :-|
all I know about the technical reasons I've tried capturing below

File: _posts/2020-03-18-how-to-use-agent-information-on-puppetserver.md (hunk 3 above)
Comment (DavidS) [reply _posts/2020-03-18-how-to-use-agent-information-on-puppetserver.md]: :+1:

File: _posts/2020-03-18-how-to-use-agent-information-on-puppetserver.md (hunk 4 above)
Comment (DavidS) [reply _posts/2020-03-18-how-to-use-agent-information-on-puppetserver.md]: Thank you :heart:

File: _posts/2020-03-18-how-to-use-agent-information-on-puppetserver.md (hunk 5 above)
Comment (DavidS) [reply _posts/2020-03-18-how-to-use-agent-information-on-puppetserver.md]: :+1:

**Pull request merged**